File: roll.pl
# Get die roll argument
my $die = @ARGV[0];

print "Rolling: ".$die."\n";

$die =~ /(?:(\d+)\s*X\s*)?(\d*)D(\d*)((?:[+\/*-]\d+)|(?:[+-][LH]))?/i
        and my ($times) = $1
        and my ($rolls) = $2
        and my ($sides) = $3
        and my ($modifier) = $4;

# If we didn't find any "times" then do it once
if ($times eq "") {
    $times = 1;
}

if ($rolls eq "") {
    $rolls = 1;
}

if ($modifier eq "") {
    $modifier = "+0";
}

# + or - if using standard die notation
$mod_operation = substr($modifier, 0, 1);
$mod_val = substr $modifier, 1;

#print "Times " . $times . "\n";
#print "Rolls " . $rolls . "\n";
#print "Sides " . $sides . "\n";
#print "Modifier " . $modifier . "\n";

$total = 0;
while ($times > 0) {

    $roll_result = 0;

    while ($rolls > 0) {

        # We are rolling the die now and adding as we go
        $roll_result += int(rand($sides)) + 1;

        $rolls--;
    }

    # Perform modifier
    if ($mod_operation eq "+") {
        $total += $roll_result + $mod_val;
    } elsif ($mod_operation eq "-") {
        $total += $roll_result - $mod_val;
    } else {
        $total += $roll_result;
    }

    $times--;
}

print $total."\n";

File: roll.lua
-- 2d6+78 format
function roll(die)

  local rolls
  local sides
  local modOperation = '+'
  local modifier = 0

  i, j = string.find(die, "d")
  if i == 1 then
    rolls = 1
  else
    rolls = tonumber(string.sub(die, 0, (j-1)))
  end

  afterD = string.sub(die, (j+1), string.len(die))
  i, j = string.find(afterD, "%d+")
  sides = tonumber(string.sub(afterD, i, j))

  afterSides = string.sub(afterD, (j+1), string.len(afterD))

  if string.len(afterSides) == 0 then
    modOperation = '+'
    modifier = 0
  else
    modOperation = string.sub(afterSides, 1, 1)
    modifier = tonumber(string.sub(afterSides, 2, string.len(afterSides)))
  end

  -- the os.time() is commonly cited as a good way to randomly
  --    seed randomness. However, die rolls aren't random in
  --    this case when running the script multiple times in one
  --    second. FOund this seeding technique on:
  --    https://www.reddit.com/r/lua/comments/1b5qg3/need_help_with_truly_random_number_generator/c940pyj/
  math.randomseed(os.time() + tonumber(tostring({}):sub(8)))

  local roll = 0
  local total = 0
  while roll < rolls do
    total = total + math.random(1, sides)
    roll = roll + 1
  end

  -- Finsihed with our rolls, now add/subtract our modifier
  if modOperation == "+" then
    total = total + modifier
  elseif modOperation == "-" then
    total = total - modifier
  end

  return total
end

die = arg[1]

if die == nil then
  print("Try:    lua roll.lue 2d6+2")
  os.exit()
end

print(roll(die))